File: scala/app/controllers/ProductsController.scala
package controllers

import javax.inject._
import play.api.mvc._
import models.ProductRepository
import play.api.data.Form
import play.api.data.Forms.mapping
import play.api.libs.json.Json
import play.api.data.Forms._

import scala.concurrent.{ExecutionContext, Future}

/**
  */
@Singleton
class ProductsController @Inject()(productRepository: ProductRepository,cc: ControllerComponents) (implicit ec: ExecutionContext )extends AbstractController(cc) {

  val productForm: Form[CreateProductForm] = Form {
    mapping(
      "name" -> nonEmptyText,
      "description"-> nonEmptyText,
      "price" -> number,
      "image" -> nonEmptyText,
      "unavailable"-> boolean,
      "categoryId" -> number
    )(CreateProductForm.apply)(CreateProductForm.unapply)
  }

  def getAll = Action.async(
    implicit request => (
      productRepository.list().map(
        product => Ok(Json.toJson(product))
      )
      )
  )

  def getById(id: Int) = Action.async { implicit request =>
    val options = for {
      maybeProduct <- productRepository.findById(id)
    } yield (maybeProduct)

    options.map { case (opt) =>
      opt match {
        case Some(product) => Ok(Json.toJson(product))
        case None => NotFound
      }
    }
  }

  def create = Action.async { implicit request =>
    productForm.bindFromRequest.fold(
      errorForm => {
        Future.successful(BadRequest("Failed to create product."))
      },
      product => {
        productRepository.create(
          product.name,
          product.description,
          product.price,
          product.image,
          product.unavailable,
          product.categoryId
        ).map { user =>
          Created(Json.toJson((user)))
        }
      }
    )
  }

  def update(id: Int) = Action { Ok("") }

  def delete(id: Int) = Action.async(
    productRepository.delete(id).map(_ => Ok(""))
  )

}

case class CreateProductForm(name: String, description: String, price: Int, image: String, unavailable: Boolean, categoryId: Int)

File: app/src/controllers/ProductsController.ts
import Axios from "axios";
import Product from "../models/Product";

type PostProductBody = {
    name: string,
    description: string,
    price: number,
    image: string | null,
    unavailable: boolean,
    categoryId: number,
}

const getAllProducts = async () => {
    try {
        const products: Product[] = (await Axios.get('/products')).data;
        return products;
    } catch (error) {
        return [];
    }
}

const getProductsForCategory = async (categoryId: number | string) => {
    try {
        const products: Product[] = (await Axios.get(`/products/byCategory/${categoryId}`)).data;
        return products;
    } catch (error) {
        return [];
    }
}

const postProduct = async (body: PostProductBody) => {
    try {
        const resp = await Axios.post(
            '/products',
            body,
            {
                headers: {
                    'Content-Type': 'application/json',
                }
            }
        )
        if (resp.status !== 201)
            return false;
        return resp.data;
    } catch (error) {
        return false;
    }
}

const deleteProduct = async (id: number | string) => {
    try {
        const resp = await Axios.delete(`/products/${id}`)
        if (resp.status !== 200)
            return false;
        return true;
    } catch (error) {
        return false;
    }
}

export { getAllProducts, getProductsForCategory, postProduct, deleteProduct, };

